### DISCUSSION TITLE
new row violates row-level security policy 'insert new rows only for authenticated users' after js client sign in

### DISCUSSION THREAD

**Discussion by pd-t** [supabase/supabase#11919] [Questions] [answered]
Hi,

I am pretty new to supabase and wanted to use the RLS to restrict access to a table, e.g.
![image](https://user-images.githubusercontent.com/67829287/214532004-83deee50-0926-467d-a1c4-d72ea9f738df.png)

So, I enabled the  RLS policy 'Enable insert access for authenticated users only' using the related quick step.
![image](https://user-images.githubusercontent.com/67829287/214527982-eb418879-3b6d-475d-a33f-465df8d832bc.png)

Now, I signed in using the js client since the API doc says:
![image](https://user-images.githubusercontent.com/67829287/214592673-1b3bdc5b-480e-4f8b-b267-aff144d16519.png)

In the end, however, I end up with the following error
![image](https://user-images.githubusercontent.com/67829287/214593368-059e5557-3f42-4266-8b66-638c69d2c995.png)

Since it is a basic and quite well documented API function, I think the error is on my side and not a bug. For this reason, I have posted the example here as a basis for discussion and hope for help.

Here is the complete code and the corresponding log (the project is already deleted - so there is no need to hide the key):
```
import { createClient } from '@supabase/supabase-js'

const supabaseurl = 'https://adlzasfknorniwzmqedf.supabase.co'
const supabasekey = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImFkbHphc2Zrbm9ybml3em1xZWRmIiwicm9sZSI6ImFub24iLCJpYXQiOjE2NzMzNTQyNjEsImV4cCI6MTk4ODkzMDI2MX0.yLk_NGkETzWv7FhTIRFDhIzcADdChIdQL7iZQ4gejjU'

var supabase = createClient(supabaseurl, supabasekey)
console.log('supabase instance: ', supabase)

const email = 'my@user.com'
const password = 'mypasswd'

await supabase.auth.signUp({
  email: email,
  password: password,
})

await supabase.auth.signInWithPassword({email: email,password: password,})
  .then(data => console.log(data))

await supabase
  .from('mytable')
  .insert({ country: 'jsland' })
  .then(data => console.log(data))
```

Error
```console
supabase instance:  SupabaseClient {
  supabaseUrl: 'https://adlzasfknorniwzmqedf.supabase.co',
  supabaseKey: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImFkbHphc2Zrbm9ybml3em1xZWRmIiwicm9sZSI6ImFub24iLCJpYXQiOjE2NzMzNTQyNjEsImV4cCI6MTk4ODkzMDI2MX0.yLk_NGkETzWv7FhTIRFDhIzcADdChIdQL7iZQ4gejjU',
  realtimeUrl: 'wss://adlzasfknorniwzmqedf.supabase.co/realtime/v1',
  authUrl: 'https://adlzasfknorniwzmqedf.supabase.co/auth/v1',
  storageUrl: 'https://adlzasfknorniwzmqedf.supabase.co/storage/v1',
  functionsUrl: 'https://adlzasfknorniwzmqedf.functions.supabase.co',
  storageKey: 'sb-adlzasfknorniwzmqedf-auth-token',
  headers: { 'X-Client-Info': 'supabase-js/2.4.0' },
  auth: SupabaseAuthClient {
    stateChangeEmitters: Map(1) { '4a931f45-6d6c-484d-94a5-00568a16873d' => [Object] },
    networkRetries: 0,
    refreshingDeferred: null,
    initializePromise: Promise { <pending> },
    detectSessionInUrl: true,
    inMemorySession: null,
    storageKey: 'sb-adlzasfknorniwzmqedf-auth-token',
    autoRefreshToken: true,
    persistSession: true,
    storage: {
      getItem: [Function: getItem],
      setItem: [Function: setItem],
      removeItem: [Function: removeItem]
    },
    admin: GoTrueAdminApi {
      url: 'https://adlzasfknorniwzmqedf.supabase.co/auth/v1',
      headers: [Object],
      fetch: [Function (anonymous)],
      mfa: [Object]
    },
    url: 'https://adlzasfknorniwzmqedf.supabase.co/auth/v1',
    headers: {
      Authorization: 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImFkbHphc2Zrbm9ybml3em1xZWRmIiwicm9sZSI6ImFub24iLCJpYXQiOjE2NzMzNTQyNjEsImV4cCI6MTk4ODkzMDI2MX0.yLk_NGkETzWv7FhTIRFDhIzcADdChIdQL7iZQ4gejjU',
      apikey: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImFkbHphc2Zrbm9ybml3em1xZWRmIiwicm9sZSI6ImFub24iLCJpYXQiOjE2NzMzNTQyNjEsImV4cCI6MTk4ODkzMDI2MX0.yLk_NGkETzWv7FhTIRFDhIzcADdChIdQL7iZQ4gejjU',
      'X-Client-Info': 'supabase-js/2.4.0'
    },
    fetch: [Function (anonymous)],
    mfa: {
      verify: [Function: bound _verify],
      enroll: [Function: bound _enroll],
      unenroll: [Function: bound _unenroll],
      challenge: [Function: bound _challenge],
      listFactors: [Function: bound _listFactors],
      challengeAndVerify: [Function: bound _challengeAndVerify],
      getAuthenticatorAssuranceLevel: [Function: bound _getAuthenticatorAssuranceLevel]
    }
  },
  fetch: [Function (anonymous)],
  realtime: RealtimeClient {
    accessToken: null,
    channels: [],
    endPoint: 'wss://adlzasfknorniwzmqedf.supabase.co/realtime/v1/websocket',
    headers: { 'X-Client-Info': 'supabase-js/2.4.0' },
    params: {
      apikey: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImFkbHphc2Zrbm9ybml3em1xZWRmIiwicm9sZSI6ImFub24iLCJpYXQiOjE2NzMzNTQyNjEsImV4cCI6MTk4ODkzMDI2MX0.yLk_NGkETzWv7FhTIRFDhIzcADdChIdQL7iZQ4gejjU'
    },
    timeout: 10000,
    transport: [Function: W3CWebSocket],
    heartbeatIntervalMs: 30000,
    heartbeatTimer: undefined,
    pendingHeartbeatRef: null,
    ref: 0,
    logger: [Function: noop],
    conn: null,
    sendBuffer: [],
    serializer: Serializer { HEADER_LENGTH: 1 },
    stateChangeCallbacks: { open: [], close: [], error: [], message: [] },
    eventsPerSecondLimitMs: 100,
    inThrottle: false,
    reconnectAfterMs: [Function (anonymous)],
    encode: [Function (anonymous)],
    decode: [Function: bound decode],
    reconnectTimer: Timer {
      callback: [Function (anonymous)],
      timerCalc: [Function (anonymous)],
      timer: undefined,
      tries: 0
    }
  },
  rest: PostgrestClient {
    url: 'https://adlzasfknorniwzmqedf.supabase.co/rest/v1',
    headers: { 'X-Client-Info': 'supabase-js/2.4.0' },
    schema: 'public',
    fetch: [Function (anonymous)]
  }
}
{
  data: {
    session: {
      access_token: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJhdXRoZW50aWNhdGVkIiwiZXhwIjoxNjc0NjQzNDMwLCJzdWIiOiI3ZGUwOTQyNS1kY2NmLTQ0YzMtYjE5OS0yMGIyYzhhNGM5YmMiLCJlbWFpbCI6Im15QHVzZXIuY29tIiwicGhvbmUiOiIiLCJhcHBfbWV0YWRhdGEiOnsicHJvdmlkZXIiOiJlbWFpbCIsInByb3ZpZGVycyI6WyJlbWFpbCJdfSwidXNlcl9tZXRhZGF0YSI6e30sInJvbGUiOiJhdXRoZW50aWNhdGVkIiwiYWFsIjoiYWFsMSIsImFtciI6W3sibWV0aG9kIjoicGFzc3dvcmQiLCJ0aW1lc3RhbXAiOjE2NzQ2Mzk4MzB9XSwic2Vzc2lvbl9pZCI6ImQxNDNlMjdjLWVhY2MtNGZhZi04NDQ5LTBjNmJkZGRiYWNlNiJ9.nBBs5fGq4AqBwfQXWNxJKlMj9QE2WHhjgY1fGb3GkBw',
      token_type: 'bearer',
      expires_in: 3600,
      refresh_token: 'hTFay4dz89CRxp8rhp2bMg',
      user: [Object],
      expires_at: 1674643431
    },
    user: {
      id: '7de09425-dccf-44c3-b199-20b2c8a4c9bc',
      aud: 'authenticated',
      role: 'authenticated',
      email: 'my@user.com',
      email_confirmed_at: '2023-01-25T09:39:56.21143Z',
      phone: '',
      confirmed_at: '2023-01-25T09:39:56.21143Z',
      last_sign_in_at: '2023-01-25T09:43:50.800076099Z',
      app_metadata: [Object],
      user_metadata: {},
      identities: [Array],
      created_at: '2023-01-25T09:39:56.204477Z',
      updated_at: '2023-01-25T09:43:50.803133Z'
    }
  },
  error: null
}
{
  error: {
    code: '42501',
    details: null,
    hint: null,
    message: 'new row violates row-level security policy for table "mytable"'
  },
  data: null,
  count: null,
  status: 401,
  statusText: 'Unauthorized'
}
```

**Comment by james-william-r** [comment]
Would strongly recommend having a look at and using the auth-helpers available. It simplifies the whole process.

https://github.com/supabase/auth-helpers

At a glance, I think you're probably sending the requests as an anonymous user due to not setting the session correctly, but the Auth helpers with help with this.

☺️

**Reply by pd-t** [reply]
Great! I will test it and give response. Meanwhile I found another discussion with similar description, but not answered yet: https://github.com/supabase/supabase/discussions/11248

**Reply by james-william-r** [reply]
Good luck! Let me know how you get on. ☺️

**Comment by pd-t** [comment]
What I forgot to say: I sent a REST call with a valid JWT token of the user with the help of the official Youtube tutorial 'Supabase Auth Deep Dive Part 3: User Based Access Policies' (https://www.youtube.com/watch?v=0LvCOlELs5U) to make sure that the RLS are set correctly server side.

This way it worked without any problems.

So the question is (maybe to someone from the supabase-js client team): Am I using the client as correctly?

**Reply by james-william-r** [reply]
So, I just read through your code properly and in answer to your question, no.

Here, you are running the database request as an anonymous user, because it's using the anonymous key inside of the createClient. After login, the key should be added to the Superbase client, then you can run the request as the user. Simply "logging in" doesn't do this.

Using the auth-helpers I linked things like this are done for you.

Make sense? ☺️

**Reply by pd-t** [reply]
Ah, this makes sense.

 The confusing thing is, that some information of the login is added to the client. For example, if I set the policy of the table to public and add a column with automatic linking to the user.id, the user id is stored correctly in the new row.

So I think the docs and the API examples are a bit confusing in this point. Does it make sense to forward this issue as ticket to the supabase-js client project? I think there should be a warning in the docs about this behaviour or the key should be added to the client automatically.

**Comment by pd-t** [comment, marked as answer]
Ok. In the meantime, I have gone a little deeper into the matter with my colleague @WograTimKoehler. To understand the whole thing you have to step back and look at the whole picture. The important information missing above is that supabase-js runs in the context of a stand-alone node.js application. Well, if you use the `signInWithPassword` method, a `getSession` call merely returns null. And this is ultimately due to the following:

``` typescript
    if (this.persistSession) {
      const maybeSession = await getItemAsync(this.storage, this.storageKey)

      if (maybeSession !== null) {
        if (this._isValidSession(maybeSession)) {
          currentSession = maybeSession
        } else {
          await this._removeSession()
        }
      }
    } else {
      currentSession = this.inMemorySession
    }
```
By default the `persistSession` setting is `true`. Unfortunately, in the node.js environment this procedure does not seem to work. Nevertheless, you can work around this by using the `inMemory` mode:

``` typescript
import { SupabaseClientOptions, createClient } from '@supabase/supabase-js'

const options: SupabaseClientOptions<any> = {auth: {persistSession: false}};
client = createClient(supabaseurl, supabasekey, options);
```

With this Workaround you get the expected behavior again. I hope this is helpful for everyone with a standalone supabase-js application. Thanks @koffietiger for your helpful remarks.
